JS: property on function "used before being assigned" regression

**TypeScript Version:**  master, due to #27128 

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:**
is used before being assigned, property on function, expando

**Code**

```js
// test.js
class TestClass {}
TestClass.NONE = 'NONE';

function fn() {
  console.log(TestClass.NONE); // error TS2565: Property 'NONE' is used before being assigned.
}
```
test with `tsc --strict --allowJs --checkJs --noEmit`

**Expected behavior:**
No error.

Error only seems to appear if the property is accessed within another function body, not top-level (or same level). Not an error without `--strictNullChecks`.

Testing with latest found this error when a property was defined on a function or class (outside its body) and then required from a different file, but it appears to be the same problem if they're in the same file, as in the example above.

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -18487,7 +18487,10 @@ namespace ts {
                     }
                 }
             }
-            else if (strictNullChecks && prop && prop.valueDeclaration && isPropertyAccessExpression(prop.valueDeclaration) && getAssignmentDeclarationPropertyAccessKind(prop.valueDeclaration)) {
+            else if (strictNullChecks && prop && prop.valueDeclaration &&
+                isPropertyAccessExpression(prop.valueDeclaration) &&
+                getAssignmentDeclarationPropertyAccessKind(prop.valueDeclaration) &&
+                getControlFlowContainer(node) === getControlFlowContainer(prop.valueDeclaration)) {
                 assumeUninitialized = true;
             }
             const flowType = getFlowTypeOfReference(node, propType, assumeUninitialized ? getOptionalType(propType) : propType);

--- a/tests/baselines/reference/typeFromPropertyAssignment37.symbols
+++ b/tests/baselines/reference/typeFromPropertyAssignment37.symbols
@@ -1,0 +1,34 @@
+=== tests/cases/conformance/salsa/use.js ===
+const util = require('./mod')
+>util : Symbol(util, Decl(use.js, 0, 5))
+>require : Symbol(require)
+>'./mod' : Symbol("tests/cases/conformance/salsa/mod", Decl(mod.js, 0, 0))
+
+function n() {
+>n : Symbol(n, Decl(use.js, 0, 29))
+
+    util.existy // no error
+>util.existy : Symbol(existy, Decl(mod.js, 1, 14))
+>util : Symbol(util, Decl(use.js, 0, 5))
+>existy : Symbol(existy, Decl(mod.js, 1, 14))
+}
+util.existy // no error
+>util.existy : Symbol(existy, Decl(mod.js, 1, 14))
+>util : Symbol(util, Decl(use.js, 0, 5))
+>existy : Symbol(existy, Decl(mod.js, 1, 14))
+
+=== tests/cases/conformance/salsa/mod.js ===
+const util = exports = module.exports = {}
+>util : Symbol(util, Decl(mod.js, 0, 5))
+>exports : Symbol("tests/cases/conformance/salsa/mod", Decl(mod.js, 0, 0))
+>module.exports : Symbol("tests/cases/conformance/salsa/mod", Decl(mod.js, 0, 0))
+>module : Symbol(module, Decl(mod.js, 0, 22))
+>exports : Symbol("tests/cases/conformance/salsa/mod", Decl(mod.js, 0, 0))
+
+if (!!false) {
+    util.existy = function () { }
+>util.existy : Symbol(existy, Decl(mod.js, 1, 14))
+>util : Symbol(util, Decl(mod.js, 0, 5))
+>existy : Symbol(existy, Decl(mod.js, 1, 14))
+}
+

--- a/tests/baselines/reference/typeFromPropertyAssignment37.types
+++ b/tests/baselines/reference/typeFromPropertyAssignment37.types
@@ -1,0 +1,44 @@
+=== tests/cases/conformance/salsa/use.js ===
+const util = require('./mod')
+>util : typeof import("tests/cases/conformance/salsa/mod")
+>require('./mod') : typeof import("tests/cases/conformance/salsa/mod")
+>require : any
+>'./mod' : "./mod"
+
+function n() {
+>n : () => void
+
+    util.existy // no error
+>util.existy : () => void
+>util : typeof import("tests/cases/conformance/salsa/mod")
+>existy : () => void
+}
+util.existy // no error
+>util.existy : () => void
+>util : typeof import("tests/cases/conformance/salsa/mod")
+>existy : () => void
+
+=== tests/cases/conformance/salsa/mod.js ===
+const util = exports = module.exports = {}
+>util : typeof import("tests/cases/conformance/salsa/mod")
+>exports = module.exports = {} : typeof import("tests/cases/conformance/salsa/mod")
+>exports : typeof import("tests/cases/conformance/salsa/mod")
+>module.exports = {} : typeof import("tests/cases/conformance/salsa/mod")
+>module.exports : typeof import("tests/cases/conformance/salsa/mod")
+>module : { "tests/cases/conformance/salsa/mod": typeof import("tests/cases/conformance/salsa/mod"); }
+>exports : typeof import("tests/cases/conformance/salsa/mod")
+>{} : {}
+
+if (!!false) {
+>!!false : boolean
+>!false : true
+>false : false
+
+    util.existy = function () { }
+>util.existy = function () { } : () => void
+>util.existy : () => void
+>util : typeof import("tests/cases/conformance/salsa/mod")
+>existy : () => void
+>function () { } : () => void
+}
+

--- a/tests/cases/conformance/salsa/typeFromPropertyAssignment37.ts
+++ b/tests/cases/conformance/salsa/typeFromPropertyAssignment37.ts
@@ -1,0 +1,15 @@
+// @noEmit: true
+// @allowJs: true
+// @checkJs: true
+// @Filename: mod.js
+const util = exports = module.exports = {}
+if (!!false) {
+    util.existy = function () { }
+}
+
+// @Filename: use.js
+const util = require('./mod')
+function n() {
+    util.existy // no error
+}
+util.existy // no error

